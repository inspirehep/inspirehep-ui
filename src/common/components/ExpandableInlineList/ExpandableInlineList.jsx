import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './ExpandableInlineList.scss';
import InlineList from '../InlineList';

class ExpandableInlineList extends Component {
  constructor(props) {
    super(props);
    this.onExpandToggle = this.onExpandToggle.bind(this);

    this.state = {
      expanded: false,
    };
  }

  onExpandToggle() {
    const { expanded } = this.state;
    this.setState({
      expanded: !expanded,
    });
  }

  renderExpandToggle() {
    const { expanded } = this.state;
    const { items } = this.props;
    const buttonText = expanded ? 'Hide' : `Show all (${items.size})`;
    return (
      /* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions, jsx-a11y/anchor-is-valid */
      <a className="toggle" onClick={this.onExpandToggle}>
        {buttonText}
      </a>
    );
  }

  render() {
    const { expanded } = this.state;
    const { limit, items, ...listProps } = this.props;
    if (!items) {
      return null;
    }

    const maybeLimitedItem = expanded ? items : items.take(limit);
    return (
      <div className="__ExpandableInlineList__">
        <InlineList items={maybeLimitedItem} {...listProps} />
        {items.size > limit && this.renderExpandToggle()}
      </div>
    );
  }
}

ExpandableInlineList.propTypes = {
  ...InlineList.propTypes,
  limit: PropTypes.number,
};

ExpandableInlineList.defaultProps = {
  ...InlineList.defaultProps,
  limit: 10,
};

export default ExpandableInlineList;
