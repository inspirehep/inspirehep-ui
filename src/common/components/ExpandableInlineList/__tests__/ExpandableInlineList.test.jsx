import React from 'react';
import { fromJS } from 'immutable';
import { shallow } from 'enzyme';

import ExpandableInlineList from '../ExpandableInlineList';

describe('ExpandableInlineList', () => {
  it('renders only 10 by default with expand toggle', () => {
    const items = fromJS([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]);
    const wrapper = shallow(
      <ExpandableInlineList items={items} renderItem={item => item} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders only limited amount with expand toggle', () => {
    const items = fromJS([1, 2, 3, 4, 5]);
    const wrapper = shallow(
      <ExpandableInlineList limit={3} items={items} renderItem={item => item} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders all onExpandToggle', () => {
    const items = fromJS([1, 2, 3, 4, 5]);
    const wrapper = shallow(
      <ExpandableInlineList limit={3} items={items} renderItem={item => item} />
    );
    const { onExpandToggle } = wrapper.instance();
    expect(wrapper.find('a')).toHaveProp('onClick', onExpandToggle);
    onExpandToggle();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
