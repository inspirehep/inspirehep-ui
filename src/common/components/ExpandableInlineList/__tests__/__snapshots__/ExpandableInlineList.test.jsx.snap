// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandableInlineList renders all onExpandToggle 1`] = `
<div
  className="__ExpandableInlineList__"
>
  <InlineList
    extractKey={[Function]}
    items={
      Immutable.List [
        1,
        2,
        3,
        4,
        5,
      ]
    }
    label={null}
    renderItem={[Function]}
    separateItems={true}
    separateItemsClassName="separate-items-with-semicolon"
    suffix={null}
    wrapperClassName={null}
  />
  <a
    className="toggle"
    onClick={[Function]}
  >
    Hide
  </a>
</div>
`;

exports[`ExpandableInlineList renders only 10 by default with expand toggle 1`] = `
<div
  className="__ExpandableInlineList__"
>
  <InlineList
    extractKey={[Function]}
    items={
      Immutable.List [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
      ]
    }
    label={null}
    renderItem={[Function]}
    separateItems={true}
    separateItemsClassName="separate-items-with-semicolon"
    suffix={null}
    wrapperClassName={null}
  />
  <a
    className="toggle"
    onClick={[Function]}
  >
    Show all (11)
  </a>
</div>
`;

exports[`ExpandableInlineList renders only limited amount with expand toggle 1`] = `
<div
  className="__ExpandableInlineList__"
>
  <InlineList
    extractKey={[Function]}
    items={
      Immutable.List [
        1,
        2,
        3,
      ]
    }
    label={null}
    renderItem={[Function]}
    separateItems={true}
    separateItemsClassName="separate-items-with-semicolon"
    suffix={null}
    wrapperClassName={null}
  />
  <a
    className="toggle"
    onClick={[Function]}
  >
    Show all (5)
  </a>
</div>
`;
